test/common/event/guarddog_impl_test.cc

```diff
@@ -0,0 +1,152 @@
+    gd.force_check();
+  }
+
+  void startNeglectedDogAndTouchedDog() {
+    InSequence s;
+    EXPECT_CALL(time_source_, currentSystemTime()).WillRepeatedly(testing::Return(time_point_));
+    GuardDogImpl gd(fakestats_, config_multikill_, time_source_);
+    auto unpet_dog = gd.getWatchDog(0);
+    auto pet_dog = gd.getWatchDog(1);
+    for (int i = 0; i < 6; i++) {
+      time_point_ += std::chrono::milliseconds(100);
+      EXPECT_CALL(time_source_, currentSystemTime()).WillRepeatedly(testing::Return(time_point_));
+      pet_dog->touch();
+      gd.force_check();
+    }
+  }
+
+  testing::NiceMock<Server::Configuration::MockMain> config_kill_;
+  testing::NiceMock<Server::Configuration::MockMain> config_multikill_;
+  testing::NiceMock<Stats::MockStore> fakestats_;
+  MockSystemTimeSource time_source_;
+  std::chrono::system_clock::time_point time_point_;
+};
+
+class GuardDogMissTest : public testing::Test {
+protected:
+  GuardDogMissTest() : config_miss(1, 1000, 0, 0), config_mega(1000, 1, 0, 0) {}
+
+  void startNeglectedDogAndWait(bool mega) {
+    EXPECT_CALL(time_source_, currentSystemTime()).WillRepeatedly(testing::Return(time_point_));
+    GuardDogImpl gd(stats_store, mega ? config_mega : config_miss, time_source_);
+    auto unpet_dog = gd.getWatchDog(0);
+    time_point_ += std::chrono::milliseconds(501);
+    EXPECT_CALL(time_source_, currentSystemTime()).WillRepeatedly(testing::Return(time_point_));
+    gd.force_check();
+    gd.stopWatching(unpet_dog);
+    unpet_dog = nullptr;
+  }
+
+  testing::NiceMock<Server::Configuration::MockMain> config_miss;
+  testing::NiceMock<Server::Configuration::MockMain> config_mega;
+  testing::NiceMock<Stats::MockStore> stats_store;
+  MockSystemTimeSource time_source_;
+  std::chrono::system_clock::time_point time_point_;
+};
+
+TEST_F(GuardDogDeathTest, KillDeathTest) { EXPECT_DEATH(startNeglectedDogAndWait(), ""); }
+TEST_F(GuardDogDeathTest, MultiKillDeathTest) { EXPECT_DEATH(startNeglectedDogsAndWait(), ""); }
+
+TEST_F(GuardDogDeathTest, NearDeathTest) {
+  startNeglectedDogAndTouchedDog();
+  SUCCEED(); // if we haven't crashed by now
+}
+
+TEST_F(GuardDogMissTest, MissTest) {
+  startNeglectedDogAndWait(false);
+  SUCCEED(); // if we haven't crashed by now
+}
+
+TEST_F(GuardDogMissTest, MegaMissTest) {
+  startNeglectedDogAndWait(true);
+  SUCCEED(); // if we haven't crashed by now
+}
+
+TEST(GuardDogBasicTest, StartStopTest) {
+  testing::NiceMock<Stats::MockStore> stats;
+  testing::NiceMock<Server::Configuration::MockMain> config(0, 0, 0, 0);
+  testing::NiceMock<MockSystemTimeSource> time_source;
+  GuardDogImpl gd(stats, config, time_source);
+}
+
+TEST(GuardDogBasicTest, LoopIntervalNoKillTest) {
+  testing::NiceMock<Stats::MockStore> stats;
+  testing::NiceMock<Server::Configuration::MockMain> config(40, 50, 0, 0);
+  testing::NiceMock<MockSystemTimeSource> time_source;
+  GuardDogImpl gd(stats, config, time_source);
+  ASSERT_EQ(gd.loopInterval(), 40);
```

Comment:
Generally best to use `EXPECT_EQ` unless your checking a condition that will make the test explode, e.g. a nullptr where later it might be used by some other test code.